### PR TITLE
add function hook changes to upgrade guide

### PR DESCRIPTION
**Pull request by kmaehashi**
update upgrade guide for changes made in #4300 and #4323.

### CODE REVIEW

File: docs/source/upgrade.rst [hunk 1]
```diff
@@ -50,6 +50,18 @@ This kind of error usually occurs by mistake (for example, not performing ``to_g
    As the argument validation is strictened, call of functions intentionally mixing NumPy/CuPy arrays in arguments will not work in Chainer v4.
    Please transfer all arrays to the same device before calling functions.
 
+Reference to Function Nodes Not Retained in Function Hooks
```
Comment (niboshi) [inline docs/source/upgrade.rst]: * "Reference" -> "References"
* This gives the impression that all function hooks are affected by this change. How about "...Not Retained in TimerHook and CupyMemoryProfilerHook"?

File: docs/source/upgrade.rst [hunk 2]
```diff
@@ -50,6 +50,18 @@ This kind of error usually occurs by mistake (for example, not performing ``to_g
...
+----------------------------------------------------------
+
+To reduce memory consumption, references to the function nodes will no longer be retained in the :class:`chainer.function_hooks.CupyMemoryProfileHook` and :class:`chainer.function_hooks.TimerHook`.
+See the discussion in `#4300 <https://github.com/chainer/chainer/pull/4300>`_ for more details.
+
+.. attention::
+
+   The existing code using function nodes retained in ``call_history`` attribute of these hooks will not work.
+   The first element of ``call_history`` became the name of the function, instead of the function node instance itself.
+   You can define own Function Hook if you need to access the function node instances.
```
Comment (niboshi) [inline docs/source/upgrade.rst]: * ...define *your* own...
* "Function Hook" -> "function hook"

**Comment by kmaehashi** [comment]
@niboshi Thanks, fixed as suggested.

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
jenkins, test this please

**Comment by niboshi** [comment]
LGTM!

**Pull request merged**